File: data-structures/avl-tree/avl-tree.js
class QNode {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class Queue {
    constructor() {
        this.first = null;
        this.last = null;
        this.size = 0;
    }
    empty() {
        return !Boolean(this.size);
    }

    enqueue(value) {
        var node = new QNode(value);
        if (!this.first) {
            this.first = node;
            this.last = this.first;
        } else {
            this.last.next = node;
            this.last = node;
        }
        this.size++;
        return this;
    }

    dequeue() {
        var currentfirst = this.first;
        if (!currentfirst) {
            return null;
        }
        this.size--;
        this.first = this.first.next;
        if (this.size === 0) {
            this.last = null;
        }
        return currentfirst.value;
    }
}

class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class AVLTree {
    constructor() {
        this.root = null;
    }

    getBalanceFactor(current=this.root) {
        return this.height(current.left) - this.height(current.right);
    }

    height(current=this.root) {
        let height = 0
        if (current === null) {
            height = -1;
        } else {
            var leftHeight = this.height(current.left);
            var rightHeight = this.height(current.right);
            height = Math.max(leftHeight, rightHeight) + 1;
        }

        return height;

    }

    rotateLL(current) {
        const pivot = current.left;
        const parent = current;
        parent.left = null
        current = pivot;
        pivot.right = parent;   
        return current;
    }

    rotateRR(current) {
        const pivot = current.right;
        const parent = current;
        parent.right = null
        current = pivot;
        pivot.left = parent;   
        return current;
    }

    rotateLR(current) {
        const pivot = current.left;
        const parent = current;
        parent.left = pivot.right;
        parent.left.left = pivot;
        pivot.right =null;
        current = this.rotateLL(current);
       return current

    }

    rotateRL(current) {
         const pivot = current.right;
        const parent = current;
        parent.right = pivot.left;
        parent.right.right = pivot;
        pivot.right = null;
        current = this.rotateRR(current)
        return current;
       
    }

    insertHelper(current, node) {
        if (current === null) {
            current = node
        } else if (node.value <= current.value) {
            current.left = this.insertHelper(current.left, node)
            if (current.left !== null && this.getBalanceFactor(current) > 1) {
                if (node.value > current.left.value) {
                    current = this.rotateLR(current);

                } else {
                    current = this.rotateLL(current)
                }
            }
        } else if (node.value > current.value) {

            current.right = this.insertHelper(current.right, node);
             if (current.right !== null && this.getBalanceFactor(current) > 1) {
                if (node.value > current.right.value) {
                    current = this.rotateRR(current);
                } else {
                    current = this.rotateRL(current)
                }
            }

        }

        return current;
    }

    insert(value) {
        const node = new Node(value);
        this.root = this.insertHelper(this.root, node)

    }

    levelOrderTraversal() {
        if (this.root === null) {
            return;
        }
        var queue = new Queue();

        queue.enqueue(this.root);
        while (!queue.empty()) {
            var node = queue.dequeue();
            console.log(node.value);
            if (node.left !== null) {
                queue.enqueue(node.left);
            }
            if (node.right !== null) {
                queue.enqueue(node.right);
            }
        }
    }

}

var avl = new AVLTree();
avl.insert(20)
avl.insert(19)
avl.insert(18)
avl.insert(17)
avl.insert(15)
avl.insert(13)
avl.insert(12)
avl.insert(18)

File: data-structures/avl-tree/avl-tree.go
package main

import (
	"fmt"
	"math"
)

type QNode struct {
	value *Node
	next  *QNode
}

type Queue struct {
	first *QNode
	last  *QNode
	size  int32
}

//pushes value at the end of the linked list
func (ll *Queue) enqueue(value *Node) {
	node := QNode{value, nil}
	if ll.first == nil {
		ll.first = &node
		ll.last = &node
	} else {
		ll.last.next = &node
		ll.last = &node

	}
	ll.size++
}

// remove first element from the list
func (ll *Queue) dequeue() *QNode {
	first := ll.first
	if first == nil {
		return nil
	}
	ll.first = ll.first.next
	if ll.size == 0 {
		ll.last = nil
	}
	ll.size--
	return first
}

func (queue *Queue) empty() bool {
	if queue.size == 0 {
		return true
	}
	return false
}

type Node struct {
	value int
	left  *Node
	right *Node
}

type AVLTree struct {
	root *Node
}

func (avlt *AVLTree) levelOrderTraversal() {
	if avlt.root == nil {
		return
	}
	queue := Queue{}
	queue.enqueue(avlt.root)
	for !queue.empty() {
		node := queue.dequeue()
		fmt.Println(node.value.value)
		if node.value.left != nil {
			queue.enqueue(node.value.left)
		}
		if node.value.right != nil {
			queue.enqueue(node.value.right)
		}

	}
}

func (avlt *AVLTree) min(current *Node) *Node {
	if current == nil {
		return nil
	}
	if current.left == nil {
		return current
	} else {
		return avlt.min(current.left)
	}
}

func (avlt *AVLTree) height(current *Node) float64 {
	if current == nil {
		return -1
	}
	leftHeight := avlt.height(current.left)
	rightHeight := avlt.height(current.right)
	return math.Max(leftHeight, rightHeight) + 1

}

func (avlt *AVLTree) rotateRight(current *Node) *Node {
	pivot := current.left
	temp := pivot.right
	pivot.right = current
	current.left=temp
	return pivot

}


func (avlt *AVLTree) rotateLeft(current *Node) *Node {
	pivot := current.right
	temp := pivot.left
	pivot.right = current
	current.right=temp
	return pivot
}

func (avlt *AVLTree) getBalanceFactor(current *Node) float64  {
	return avlt.height(current.left) - avlt.height(current.right)
}

func (avlt *AVLTree) insert(value int) {
	node := Node{value, nil, nil}
	avlt.root = avlt.insertHelper(avlt.root, &node)

}
func (avlt *AVLTree) insertHelper(current *Node, node *Node) *Node {
	if current == nil {
		current = node
	}else if node.value <= current.value {
		current.left = avlt.insertHelper(current.left,node)
		if current.left != nil && avlt.getBalanceFactor(current) > 1{
			if node.value > current.left.value {
				current = avlt.rotateRight(current)
				current = avlt.rotateLeft(current)
			} else {
				current = avlt.rotateRight(current)
			}
		}
	}else if node.value > current.value {
		current.right = avlt.insertHelper(current.right,node)
		if current.right != nil && avlt.getBalanceFactor(current) > 1 {
			if node.value > current.right.value {
				current = avlt.rotateLeft(current);
			} else {
				current = avlt.rotateLeft(current)
				current = avlt.rotateRight(current)

			}
		}
	}
	return current

	// if avlt.root == nil {
	// 	node := Node{value, nil, nil}
	// 	avlt.root = &node
	// 	return avlt
	// } else {
	// 	current := avlt.root
	// 	for {
	// 		if value <= current.value {
	// 			if current.left == nil {
	// 				node := Node{value, nil, nil}
	// 				current.left = &node
	// 				return avlt
	// 			} else {
	// 				current = current.left
	// 			}
	// 		} else {
	// 			if current.right == nil {
	// 				node := Node{value, nil, nil}
	// 				current.right = &node
	// 				return avlt
	// 			} else {
	// 				current = current.right
	// 			}
	// 		}
	// 	}
	// }
}
func (avlt *AVLTree) delete(value int)  {
	avlt.root = avlt.deleteHelper(avlt.root,value)
}
func (avlt *AVLTree) deleteHelper(current *Node, value int) *Node {
	if current == nil {
		return current
	} else if value < current.value {
		current.left = avlt.deleteHelper(current.left, value)
	} else if value > current.value {
		current.right = avlt.deleteHelper(current.right, value)
	} else {
		if current.left == nil && current.right == nil {
			current = nil
		} else if current.left == nil {
			current = current.right
		} else if current.right == nil {
			current = current.left
		} else {
			minRight := avlt.min(current.right)
			current.value = minRight.value
			current.right = avlt.deleteHelper(current.right, minRight.value)
		}
	}

	if current!= nil {
		 balanceFactor := avlt.getBalanceFactor(current)

		if balanceFactor > 1 && avlt.getBalanceFactor(current.left) >= 0 {
			current = avlt.rotateRight(current)
		}
		if balanceFactor > 1 && avlt.getBalanceFactor(current.left) < 0 {
			current = avlt.rotateRight(current)
			current = avlt.rotateLeft(current)
		}
		if balanceFactor < -1 && avlt.getBalanceFactor(current.right) <= 0 {
			current = avlt.rotateLeft(current)
		}
		if balanceFactor < -1 && avlt.getBalanceFactor(current.right) > 0 {
			current = avlt.rotateLeft(current)
			current = avlt.rotateRight(current)
		}
	}
	return current
}

func main() {
	avlt := AVLTree{nil}
	avlt.insert(30)
	avlt.insert(10)
	avlt.insert(5)
	avlt.insert(3)
	avlt.delete(30)
	avlt.levelOrderTraversal()
}
